R/description.R

```diff
@@ -51,63 +51,59 @@
-  tidy_desc(desc)
+  desc <- build_description(fields, package = name)
   lines <- desc$str(by_field = TRUE, normalize = FALSE, mode = "file")
 
   write_over(proj_path("DESCRIPTION"), lines)
-
   if (!getOption("usethis.quiet", default = FALSE)) {
     print(desc)
   }
 }
 
 #' @rdname use_description
+#' @param package Package name
 #' @export
-use_description_defaults <- function() {
-  list(
-    usethis.description = getOption("usethis.description"),
-    devtools.desc = getOption("devtools.desc"),
-    usethis = list(
-      Package = "valid.package.name.goes.here",
-      Version = "0.0.0.9000",
-      Title = "What the Package Does (One Line, Title Case)",
-      Description = "What the package does (one paragraph).",
-      "Authors@R" = 'person("First", "Last", , "first.last@example.com", c("aut", "cre"), comment = c(ORCID = "YOUR-ORCID-ID"))',
-      License = " `use_mit_license()`, `use_gpl3_license()` or friends to pick a license",
-      Encoding = "UTF-8",
-      LazyData = "true",
-      Roxygen = "list(markdown = TRUE)"
-    )
+use_description_defaults <- function(package = NULL, fields = list()) {
+  check_is_named_list(fields)
+
+  usethis <- list(
+    Package = package %||% "valid.package.name.goes.here",
+    Version = "0.0.0.9000",
+    Title = "What the Package Does (One Line, Title Case)",
+    Description = "What the package does (one paragraph).",
+    "Authors@R" = 'person("First", "Last", , "first.last@example.com", c("aut", "cre"), comment = c(ORCID = "YOUR-ORCID-ID"))',
+    License = "`use_mit_license()`, `use_gpl3_license()` or friends to pick a license",
+    Encoding = "UTF-8",
+    LazyData = "true",
+    Roxygen = "list(markdown = TRUE)"
   )
-}
+  options <- getOption("usethis.description") %||% getOption("devtools.desc") %||% list()
 
-build_description <- function(fields = list()) {
-  desc_list <- build_description_list(fields)
+  defaults <- utils::modifyList(usethis, options)
+  defaults <- utils::modifyList(defaults, fields)
 
-  # Collapse all vector arguments to single strings
-  desc <- vapply(desc_list, glue_collapse, character(1))
+  # Ensure each element is a single string
+  if (inherits(defaults$`Authors@R`, "person")) {
+    defaults$`Authors@R` <- format(defaults$`Authors@R`, style = "R")
+    defaults$`Authors@R` <- paste0(defaults$`Authors@R`, collapse = "\n")
+  }
+  defaults <- lapply(defaults, paste, collapse = "")
 
-  glue("{names(desc)}: {desc}")
+  compact(defaults)
 }
 
-build_description_list <- function(fields = list()) {
-  defaults <- use_description_defaults()
-  defaults <- utils::modifyList(
-    defaults$usethis,
-    defaults$usethis.description %||% defaults$devtools.desc %||% list()
-  )
-  compact(utils::modifyList(defaults, fields))
+build_description <- function(fields, package) {
+  fields <- use_description_defaults(package, fields)
+
+  desc <- desc::desc(text = glue("{names(fields)}: {fields}"))
+  tidy_desc(desc)
+  desc
```

Comment:
This looks odd. Maybe I don't remember what everything does, but worth a check.

Reply:
It looks weird, but it modifies the mutable desc object in place.

Reply:
OK right I did not recall the R6-ness of `desc`.